exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,618 @@
+  /// https://sparkplug.eclipse.org/specification/version/3.0/documents/sparkplug-specification-3.0.0.pdf
+  auto type_enum_convert(tfc::ipc::details::type_e type) -> uint32_t {
+    switch (type) {
+      case details::type_e::unknown:
+        return 0;
+      case tfc::ipc::details::type_e::_bool:
+        return 11;
+      case details::type_e::_int64_t:
+        return 4;
+      case details::type_e::_uint64_t:
+        return 8;
+      case details::type_e::_double_t:
+        return 10;
+      case details::type_e::_string:
+      case details::type_e::_json:
+        return 12;
+    }
+    return 0;
+  }
+
+  auto format_signal_name(std::string signal_name_to_format) -> std::string {
+    std::replace(signal_name_to_format.begin(), signal_name_to_format.end(), '.', '/');
+    return signal_name_to_format;
+  }
+
+  auto topic_formatter(const std::vector<std::string_view>& topic_vector) -> std::string {
+    if (topic_vector.empty()) {
+      throw std::runtime_error("Topic can not be empty");
+    }
+    std::string topic;
+    for (const auto& sub_topic : topic_vector) {
+      topic += std::string(sub_topic) + "/";
+    }
+    topic.pop_back();
+    return topic;
+  }
+
+  auto set_value_payload(Payload_Metric* metric, std::any value) -> void {
+    if (value.type() == typeid(bool)) {
+      metric->set_boolean_value(std::any_cast<bool>(value));
+    } else if (value.type() == typeid(std::string)) {
+      metric->set_string_value(std::any_cast<std::string>(value));
+    } else if (value.type() == typeid(uint64_t)) {
+      metric->set_long_value(std::any_cast<uint64_t>(value));
+    } else if (value.type() == typeid(int64_t)) {
+      metric->set_long_value(std::any_cast<int64_t>(value));
+    } else if (value.type() == typeid(double)) {
+      metric->set_double_value(std::any_cast<double>(value));
+    } else if (value.type() == typeid(float)) {
+      metric->set_float_value(std::any_cast<float>(value));
+    } else if (value.type() == typeid(uint32_t)) {
+      metric->set_int_value(std::any_cast<uint32_t>(value));
+    } else {
+      throw std::runtime_error("Unexpected type in std::any.");
+    }
+  }
+
+  auto set_value_payload(Payload_Metric* metric, bool& value) -> void { metric->set_boolean_value(value); }
+
+  auto set_value_payload(Payload_Metric* metric, std::string& value) -> void { metric->set_string_value(value); }
+
+  auto set_value_payload(Payload_Metric* metric, uint64_t& value) -> void { metric->set_long_value(value); }
+
+  auto set_value_payload(Payload_Metric* metric, int64_t& value) -> void { metric->set_long_value(value); }
+
+  auto set_value_payload(Payload_Metric* metric, double& value) -> void { metric->set_double_value(value); }
+
+  auto set_value_payload(Payload_Metric* metric, float& value) -> void { metric->set_float_value(value); }
+
+  auto set_value_payload(Payload_Metric* metric, uint32_t& value) -> void { metric->set_int_value(value); }
+
+  auto connect_to_broker() -> asio::awaitable<void> {
```

Comment:
I would like to see you clean up this method.
try not using throw catch as goto's.

You can use asio::as_tuple to get rid of the exceptions and
get a cleaner control flow inside the method. You can also compose some async operations
like "connect_and_handshake" and hide the implementation details away from
the reconnecting logic.

I would also say some of these operations dont belong inside the loop. Why do we have to set_default_verify_paths for the ssl context each time a connect fails? or the verify mode.

This could also be seen to be as broad because you dont consider which operation failed.

Reply:
I refactored the function into two subfunctions and changed the control flow. Please review the new function when possible.